wolvic/browser/vr/wvr_manager.cc

```diff
@@ -502,12 +502,75 @@ void WvrManager::DrawFrameSubmitNow(device::WebXrFrame* processing_frame) {
   // Renderer is waiting for the previous frame to render, unblock it now.
   submit_client_->OnSubmitFrameRendered();
 
-  if (webxr_.HaveRenderingFrame())
-    webxr_.EndFrameRendering();
+  if (webxr_.HaveRenderingFrame()) {
+    // It's possible, though unlikely, that the previous rendering frame hasn't
+    // finished yet, for example if an unusually slow frame is followed by an
+    // unusually quick one. In that case, wait for that frame to finish
+    // rendering first before proceeding with this one. The state machine
+    // doesn't permit two frames to be in rendering state at once. (Also, adding
+    // even more GPU work in that condition would be counterproductive.)
+    DVLOG(2) << __func__ << ": wait for previous rendering frame to complete";
+    FinishRenderingFrame();
+  }
+
+  CHECK(!webxr_.HaveRenderingFrame());
+  CHECK(webxr_.HaveProcessingFrame());
+  auto* frame = webxr_.GetProcessingFrame();
+  frame->time_copied = base::TimeTicks::Now();
+
+  frame->render_completion_fence = nullptr;
   webxr_.TransitionFrameProcessingToRendering();
 
-  // See if we can animate a new WebXR frame.
-  WebXrTryStartAnimatingFrame();
+  // We finished processing a frame, unblock a potentially waiting next frame.
+  webxr_.TryDeferredProcessing();
```

Comment:
In which thread is this happening?

It would be good to add thread or sequence checks so we make sure things are happening at the thread we expect.

And, about this case, TryDeferredProcessing, can it take too much time? Is it blocking a thread that should not be blocked for long?

Unfortunately just reviewing I am not sure where this is happening.